Implement bswap in SAT back-end

tautschnig opened the pull request:
Previously only constants were supported via expression simplification.

tautschnig committed:
Implement bswap in SAT back-end

Previously only constants were supported via expression simplification.

smowton commented on regression/cbmc/gcc_bswap1/main.c:
```diff
@@ -0,0 +1,26 @@
+  assert(__builtin_bswap32(a)==0x00efcdab);
+  unsigned b;
+  b&=0xffffff00;
+  assert((__builtin_bswap32(b)&0xff000000)==0);
```
Note GCC has bswap64 as well, might as well get that while we're here?

smowton commented on src/solvers/flattening/boolbv_bswap.cpp:
```diff
@@ -0,0 +1,43 @@
+  std::size_t width=boolbv_width(expr.type());
+
+  // width must be multiple of bytes
+  if(width%config.ansi_c.char_width!=0)
```
Now that operator spacing is allowed in the CBMC codebase, let's use it to improve readability :)

tautschnig commented:
@smowton Thanks a lot for the comments, all addressed.

romainbrenguier commented on src/solvers/flattening/boolbv_bswap.cpp:
```diff
@@ -0,0 +1,43 @@
+
+bvt boolbvt::convert_bswap(const exprt &expr)
+{
+  if(expr.id() != ID_bswap || expr.operands().size() != 1)
```
Having a precondition seems more appropriate

romainbrenguier replied:
I don't like that we can have expressions with their own id and no associated class (inheriting from `exprt`). In particular it is not documented anywhere how these bswap expressions are structured.

romainbrenguier commented on src/solvers/flattening/boolbv_bswap.cpp:
```diff
@@ -0,0 +1,43 @@
+  if(expr.id() != ID_bswap || expr.operands().size() != 1)
+    throw "expected bswap with one operand";
+
+  std::size_t width = boolbv_width(expr.type());
```
should be const

romainbrenguier commented on src/solvers/flattening/boolbv_bswap.cpp:
```diff
@@ -0,0 +1,43 @@
+/*******************************************************************\
+
+Module:
```
I think we should have a module name or description here

romainbrenguier commented on regression/cbmc/gcc_bswap1/test.desc:
```diff
@@ -0,0 +1,8 @@
+
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
```
Having one positive test does not seem enough to show that it works (a contradiction in the solver would give the same result).
We should have also some tests for which verification fail (for instance bswap a particular value and assert it is different from the expected result).

tautschnig committed:
Introduce bswap_exprt

ID_bswap was previously used without a dedicated class.

tautschnig commented:
@romainbrenguier Thank you for pushing me to do the cleanup! I believe I have addressed all the issues you raised.

martin-cs commented on regression/cbmc/gcc_bswap1/main.c:
```diff
@@ -0,0 +1,37 @@
+{
+  uint16_t sa = 0xabcd;
+  assert(__builtin_bswap16(sa) == 0xcdab);
+  uint16_t sb;
```
Pedantry but, no chance of a nondet_whatever() function call for the things that should be nondet.  Doesn't change behaviour but helps people see what is going on.

martin-cs commented on src/solvers/flattening/boolbv_bswap.cpp:
```diff
@@ -0,0 +1,41 @@
+    return conversion_failed(expr);
+
+  bvt result = convert_bv(expr.op());
+  CHECK_RETURN(result.size() == width);
```
Tidy!  I also note that convert_bv is correct, convert_bitvector is not.

martin-cs commented on src/util/std_expr.h:
```diff
@@ -442,6 +442,78 @@ inline void validate_expr(const unary_minus_exprt &value)
+  DATA_INVARIANT(
+    expr.op0().type()==expr.type(),
+    "bswap type must match operand type");
+  return static_cast<const bswap_exprt &>(expr);
```
DATA_INVARIANTs about types are a most welcome addition.  @chrisr-diffblue will approve!

martin-cs reviewed (approved):
LGTM.  Am willing to merge once @romainbrenguier says he is happy with the changes he requested.

tautschnig commented:
@romainbrenguier @martin-cs I believe I've addressed all change suggestions (unless I misunderstood any of the above comments).

romainbrenguier commented on regression/cbmc/gcc_bswap1/test.desc:
```diff
@@ -0,0 +1,11 @@
+^EXIT=10$
+^SIGNAL=0$
+^VERIFICATION FAILED$
+\[main\.assertion\..\] assertion __builtin_bswap(16|32|64)\((sb|b|bl)\) == 0: FAILURE$
```
could you put the assertion number so that it is clear which ones are supposed to fail

tautschnig replied:
Done.

tautschnig commented:
Pinging @kroening for a review.

tautschnig commented:
@kroening ping?

peterschrammel reviewed (approved):
Lgtm

tautschnig commented:
Will still need @kroening's approval...

hannes-steffenhagen-diffblue commented on regression/cbmc/gcc_bswap1/main.c:
```diff
@@ -0,0 +1,41 @@
+  uint16_t sb = __VERIFIER_nondet_u16();
+  sb &= 0xff00;
+  // expected to fail, only MSB guaranteed to be 0
+  assert(__builtin_bswap16(sb) == 0);
```
Shouldn't this be unknown instead of failing if it's unknown?

tautschnig replied:
Is that asking to clarify the comment (which should likely say "may fail")?

tautschnig committed:
C library: network byteorder functions

On some systems htonl etc. are just macros (always true for big-endian systems,
where they are null macros; and also true on Mac OS). For others we to provide
the implementation. In all cases the full support for bswap* is necessary, and
that is now available as of the preceding commit.

Also fixes arpa/inet.h being unavailable on Windows/Visual Studio.
